### PR TITLE
Pythia8 for high energy modules

### PR DESCRIPTION
**Pull request by pochoarus**
Changes to be able to run pythia8 with high energy modules.

LeptoHadronization and PhotonCOHGenerator will use the Singleton. GLRESGenerator and PhotonRESGenerator have its own pythia8 class.

### CODE REVIEW

**Review by nusense — COMMENTED** [review]
One thing I think I question about how this is structured is that I think it's all going to go "wrong" if both pythia6 and pythia8 are enabled at config/build time.   For the baseline code doing this is an acceptable thing to because the choice of which to use at runtime is set by configurations in the XML files.   Here I think what happens is both the pythia6 and pythia8 parts of the code get run in sequence to uncertain ends.

File: src/Physics/HELepton/EventGen/GLRESGenerator.cxx [hunk 1]
```diff
@@ -246,15 +274,83 @@ void GLRESGenerator::ProcessEventRecord(GHepRecord *
...
+    fPythia->next();
+
+    // fPythia->event.list();
+    // fPythia->stat();
+
+    Pythia8::Event &fEvent = fPythia->event;
+    int np = fEvent.size();
+    assert(np>0);
+
+    for (int i = 5; i < np; i++) { // ignore firt entry -> system + input particles
```
[the remark this answers is not in the document]
Comment (pochoarus) [reply src/Physics/HELepton/EventGen/GLRESGenerator.cxx:297]: done

File: src/Physics/HELepton/EventGen/GLRESGenerator.cxx [hunk 2]
```diff
@@ -246,15 +274,83 @@ void GLRESGenerator::ProcessEventRecord(GHepRecord *
...
+
+      int pdgc = fEvent[i].id();
+      if (!PDGLibrary::Instance()->Find(pdgc)) continue; // some intermidatie particles not part of genie tables
```
[the remark this answers is not in the document]
Comment (pochoarus) [reply src/Physics/HELepton/EventGen/GLRESGenerator.cxx:300]: done

File: src/Physics/HELepton/EventGen/Makefile [hunk 3]
```diff
@@ -25,4 +25,20 @@ install : install-inc install-lib
 #
 include $(GENIE)/src/make/Make.std-package-targets
 
+ifeq ($(strip $(GOPT_ENABLE_PYTHIA6)),YES)
+  CPP_INCLUDES           += $(PYTHIA6_INCLUDES)
+  ROOT_DICT_GEN_INCLUDES += $(PYTHIA6_INCLUDES)
+  EXTRA_EXT_LIBS         += $(PYTHIA6_LIBRARIES)
+else
+  $(info $(PACKAGE) not build against Pythia6 )
```
[the remark this answers is not in the document]
Comment (pochoarus) [reply src/Physics/HELepton/EventGen/Makefile:33]: done

File: src/Physics/HEDIS/EventGen/Makefile
[the remark this answers is not in the document]
Comment (pochoarus) [reply src/Physics/HEDIS/EventGen/Makefile:1]: i have also fixed it in Decay and Hadronization Makefiles

File: src/Physics/HELepton/EventGen/PhotonCOHGenerator.cxx [hunk 4]
```diff
@@ -216,11 +227,10 @@ void PhotonCOHGenerator::ProcessEventRecord(GHepRecord * evrec) const
 
     }
 
-    double lightspeed = 299792458e3; //c in mm/s. Used for time in PYTHIA t[s]=t_pythia[mm]/c[mm/s]
     double vx = nu->X4()->X() + particle->GetVx()*1e12; //pythia gives position in [mm] while genie uses [fm]
     double vy = nu->X4()->Y() + particle->GetVy()*1e12;
     double vz = nu->X4()->Z() + particle->GetVz()*1e12;
-    double vt = nu->X4()->T() + particle->GetTime()/lightspeed;
+    double vt = nu->X4()->T() + particle->GetTime()*(units::millimeter/units::second);
```
[the remark this answers is not in the document]
Comment (pochoarus) [reply src/Physics/HELepton/EventGen/PhotonCOHGenerator.cxx:233]: see above

File: src/Physics/HELepton/EventGen/PhotonCOHGenerator.h [hunk 5]
```diff
@@ -54,6 +60,11 @@ public :
   mutable TPythia6 * fPythia;   ///< PYTHIA6 wrapper class
 #endif
 
+#ifdef __GENIE_PYTHIA8_ENABLED__
+  mutable Pythia8::Pythia * fPythia;         ///< PYTHIA6 wrapper class
```
[the remark this answers is not in the document]
Comment (pochoarus) [reply src/Physics/HELepton/EventGen/PhotonCOHGenerator.h:64]: done

File: src/Physics/HELepton/EventGen/PhotonRESGenerator.cxx [hunk 6]
```diff
@@ -11,7 +11,6 @@
 #include <cstring>
 
 #include <RVersion.h>
-#include "Framework/Conventions/GBuild.h"
```
[the remark this answers is not in the document]
Comment (pochoarus) [reply src/Physics/HELepton/EventGen/PhotonRESGenerator.cxx:14]: The thing is that I have it also defined in PhotonRESGenerator.h. That is why I remove it from the .cxx file, to avoid duplication.

File: src/Physics/HELepton/EventGen/PhotonRESGenerator.h [hunk 7]
```diff
@@ -54,6 +58,12 @@ public :
   mutable TPythia6 * fPythia;   ///< PYTHIA6 wrapper class
 #endif
 
+#ifdef __GENIE_PYTHIA8_ENABLED__
+  // we need to classes because we have to simulate anue+e->W- (N) and nue+e+>W+ (P) decays
```
[the remark this answers is not in the document]
Comment (pochoarus) [reply src/Physics/HELepton/EventGen/PhotonRESGenerator.h:62]: done

**Comment by pochoarus** [comment]
Ok. I am going to change the code to follow the same structure of Pythia*Hadro2019. One base class and split pythia6 and pythia8 codes. The user will have to define which one to use in the config file.

**Comment by pochoarus** [comment]
The new version uses independent classes for pythia6 and pythia8. The user can choose between them in the configuration files.

File: src/Physics/HELepton/EventGen/PhotonRESWdecPythia6.cxx [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+      double vz = nu->X4()->Z() + particle->GetVz()*1e12;
+      double vt = nu->X4()->T() + particle->GetTime()/lightspeed;
+      TLorentzVector pos( vx, vy, vz, vt );
+
+      event->AddParticle(pdgc, ist, firstmother, lastmother, firstchild, lastchild, p4o, pos );
+
+    }
+
+    delete particle;
+    pythia_particles->Clear("C");
```
Comment (nusense) [inline src/Physics/HELepton/EventGen/PhotonRESWdecPythia6.cxx:244]: Just checking here.  The `delete particle` would delete just the last entry of the `pythia_particles` list, would it not?  I'm not entirely clear on the ownership model behind all this, but would like assurances that the call to `Clear("C")` doesn't cause a double deletion.  Because there's nothing intervening, I'm not sure anything bad would happen if that were the case, but it best not to have that happen.   The alternative possibility is that it should be deleting each particle as it's iterating and we have a potential memory leak.  I'm not sure what the `Clear()` is actually doing.

File: src/Physics/Hadronization/LeptoHadPythia6.cxx [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+        pyrobo_( &imin , &imax, &the, &ph, &dbex, &dbey , &dbez );
+        double theta_hadron = pyangl_(&pz_hadron,&pT);
+
+        fPythia->SetMSTU( 10, 1 ); //keep the mass value stored in P(I,5), whatever it is.
+        fPythia->SetP( fPythia->GetN()+1, 5, m_hadron );
+        fPythia->Py1ent( fPythia->GetN()+1, hadron, E_hadron, theta_hadron, phi + kPi );
+        fPythia->SetMSTU( 10, 2 ); //find masses according to mass tables as usual.
+
+        // Target remnants required to go backwards in hadronic cms
+        if ( fPythia->GetP(fPythia->GetN()-1,3)<0 && fPythia->GetP(fPythia->GetN(),3)<0 ) break; //quit the while from line 368
```
Comment (nusense) [inline src/Physics/Hadronization/LeptoHadPythia6.cxx:293]: This combination of while loops and if's is a bit hard to follow, I'll assume this is right.   But I'm guessing that the comment after the `break` isn't right about line number any more.  I'd suggest that rather than using the line number it give the actual while conditional.

**Review by nusense — APPROVED** [review]
Looking over things there are a couple of quibbles that might be addressed, but I don't think are showstoppers at this time.  These latest changes do seem to address my concern about if GENIE is built with both pythia6 and pythia8 enabled simultaneously.  I'm going to approve so we can move forward.

**Review by nusense — APPROVED** [review]
I'm not sure why my previous set of comments didn't trigger the approved status, but I'll try again here

**Comment by sjgardiner** [comment]
Thanks, both, I will go ahead and merge.
